### PR TITLE
ConvectiveAdjustmentVerticalDiffusivity

### PR DESCRIPTION
**Pull request by glwagner**
This PR implements a relatively simple convective adjustment scheme that provides the option for convective adjustment of both momentum and tracers. The closure scheme implements only a vertical diffusivity, and has a convenience "background diffusivity" in addition to a convective diffusivity that is active when the buoyancy gradient is unstable. Both the convective and background diffusivities may be functions of space and time, or may be discrete fields, and can be different for each tracer.

Here's a result from a new validation test:

image

I also need to add unit tests.

### CODE REVIEW

File: src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    @inbounds  diffusivities.stable_buoyancy_gradient[i, j, k] = is_stableᶜᶜᶠ(i, j, k, grid, tracers, buoyancy)
+end
+
+#####
+##### Fluxes
+#####
+
+# u
+@inline function νᶠᶜᶠ(i, j, k, grid, clock, closure::CAVD, stable_buoyancy_gradient)
+    @inbounds stableᶠᶜᶠ = stable_buoyancy_gradient[i, j, k] || stable_buoyancy_gradient[i+1, j, k]
```
Comment (christophernhill) [inline src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl:100]: @glwagner is ```i+1``` correct? Isn't ```u[i,j,k]``` bracketed by ```rho[i-1,j,k]``` and ```rho[i,j,k]```?

File: src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+end
+
+@inline function viscous_flux_uz(i, j, k, grid, closure::CAVD, clock, velocities, diffusivities, tracers, buoyancy)
+    ν = νᶠᶜᶠ(i, j, k, grid, clock, closure, diffusivities.stable_buoyancy_gradient)
+    return - ν * ∂zᵃᵃᶠ(i, j, k, grid, velocities.u)
+end
+
+# v
+@inline function νᶜᶠᶠ(i, j, k, grid, clock, closure::CAVD, stable_buoyancy_gradient)
+    @inbounds stableᶜᶠᶠ = stable_buoyancy_gradient[i, j, k] || stable_buoyancy_gradient[i, j+1, k]
```
Comment (christophernhill) [inline src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl:116]: same here ?

File: src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+end
+
+@inline function viscous_flux_vz(i, j, k, grid, closure::CAVD, clock, velocities, diffusivities, tracers, buoyancy)
+    ν = νᶜᶠᶠ(i, j, k, grid, clock, closure, diffusivities.stable_buoyancy_gradient)
+    return - ν * ∂zᵃᵃᶠ(i, j, k, grid, velocities.v)
+end
+
+# w
+@inline function νᶜᶜᶜ(i, j, k, grid, clock, closure::CAVD, stable_buoyancy_gradient)
+    @inbounds stableᶜᶜᶜ = stable_buoyancy_gradient[i, j, k] || stable_buoyancy_gradient[i, j, k+1]
```
Comment (christophernhill) [inline src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl:132]: and here?

Comment (glwagner) [reply src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl:132]: Now this one I think is right. But on the others, I agree with you!

File: src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl (hunk 1 above)
Comment (glwagner) [reply src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl:100]: quite, good eye 👁️  !

File: src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl (hunk 2 above)
Comment (glwagner) [reply src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl:116]: yes

File: src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl (hunk 3 above)
Comment (glwagner) [reply src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl:132]: to explain: for the others we evaluate an object at centers to faces in i or j respectively. here, we evaluate an object located at faces to cell centers --- the reverse logic

File: src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    @inbounds  diffusivities.stable_buoyancy_gradient[i, j, k] = is_stableᶜᶜᶠ(i, j, k, grid, tracers, buoyancy)
+end
+
+#####
+##### Fluxes
+#####
+
+# u
+@inline function νᶠᶜᶠ(i, j, k, grid, clock, closure::CAVD, stable_buoyancy_gradient)
+    @inbounds stableᶠᶜᶠ = stable_buoyancy_gradient[i, j, k] || stable_buoyancy_gradient[i-1, j, k]
```
Comment (navidcy) [inline src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl:100]: I don't understand this line, why either `i` or `i-1` and not also `i+1`?
probably something that has to do with staggered grid details that I'm missing..

**Review by navidcy — COMMENTED** [review]
I can't really follow _all_ code details. Some questions/remarks:

- Does this work with VerticallyStretched grids supposedly? If not, then add a remark at the docstring?
- I'm trying to think perhaps a bit more elaborate validation experiment to demonstrate this functionality. Perhaps an initial buoyancy profile that has both stable and unstable regions so that we can see that there is different behavior in each?

File: src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl (hunk 4 above)
Comment (glwagner) [reply src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl:100]: The buoyancy gradient is defined at `Center, Center, Face` (superscript ccf). Thus we determine that the node fcf at Face, Center, Face is stable by checking the stability of the buoyancy gradient on either side of the node. By convention, face indices lie to the "left" of centered indices. Thus the buoyancy gradients at i and i-1 "span" the node at fcf.

**Comment by glwagner** [comment]
> Does this work with VerticallyStretched grids supposedly? If not, then add a remark at the docstring?

This works with all grids. All closures work on vertically stretched grids --- however there are a few (the LES closures) that don't work on curvilinear grids. Perhaps it'd be better to add comments about the restrictions in those closures? That's for a separate PR / issue though.

File: src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl (hunk 4 above)
Comment (navidcy) [reply src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl:100]: OK, makes sense. I was missing this "convention".

**Comment by navidcy** [comment]
@glwagner, do you think a bit more elaborate validation test is in order?
I'll approve at this point and let this to your judgment.

**Review by navidcy — APPROVED** [review]

**Comment by glwagner** [comment]
> @glwagner, do you think a bit more elaborate validation test is in order?
> I'll approve at this point and let this to your judgment.

I am currently working on a more elaborate validation / example (I'd like to implement an overturning, eddying channel example that uses this closure in a 3D setting) so I think we are ok to merge this PR now as is.

File: src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl (hunk 4 above)
Comment (glwagner) [reply src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl:100]: This is a modeling choice --- I'm not aware that such an implementation of convective adjustment is standard or not. In fact, most convective adjustment schemes may not diffuse the velocity field (only tracers).

Comment (navidcy) [reply src/TurbulenceClosures/turbulence_closure_implementations/convective_adjustment_vertical_diffusivity.jl:100]: I meant the convention regarding "face indices lie to the "left" of centered indices".

**Comment by navidcy** [comment]
> > @glwagner, do you think a bit more elaborate validation test is in order?
> > I'll approve at this point and let this to your judgment.
> 
> I am currently working on a more elaborate validation / example (I'd like to implement an overturning, eddying channel example that uses this closure in a 3D setting) so I think we are ok to merge this PR now as is.

Let's work on the eddying channel together!

**Pull request merged**